File: prototype/rails/src/test/domain/shop/checkout_purchaser_test.rb
require "test_helper"

class CheckoutPurchaserTest < ActiveSupport::TestCase
  CheckoutPurchaser = Domain::Shop::CheckoutPurchaser

  test "a guest buys under the address they typed" do
    purchaser = CheckoutPurchaser.for_checkout(
      id: 7, account_email: nil, account_verified_at: nil, submitted_email: "  Ada@Example.Test "
    )

    assert_equal 7, purchaser.id
    assert_equal "ada@example.test", purchaser.email
    refute purchaser.email_verified?
  end

  test "a signed in customer buys under the address on their account" do
    verified_at = Time.utc(2026, 8, 20, 9)

    purchaser = CheckoutPurchaser.for_checkout(
      id: 7, account_email: "ada@example.test", account_verified_at: verified_at,
      submitted_email: "someone-else@example.test"
    )

    assert_equal "ada@example.test", purchaser.email
    assert_equal verified_at, purchaser.email_verified_at
    assert purchaser.email_verified?
  end
end

File: prototype/node/src/app/core/shop/checkout-purchaser.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { purchaserForCheckout } from './checkout-purchaser.ts'

test('a guest buys under the address they typed', () => {
  const purchaser = purchaserForCheckout({
    customerId: 7,
    accountEmail: null,
    isAccountVerified: false,
    submittedEmail: '  Ada@Example.Test ',
  })

  assert.deepEqual(purchaser, { id: 7, email: 'ada@example.test', isEmailVerified: false })
})

test('a signed-in customer buys under the address on their account', () => {
  const purchaser = purchaserForCheckout({
    customerId: 7,
    accountEmail: 'ada@example.test',
    isAccountVerified: true,
    submittedEmail: 'someone-else@example.test',
  })

  assert.deepEqual(purchaser, { id: 7, email: 'ada@example.test', isEmailVerified: true })
})

test('an account with no address on it still buys as a guest', () => {
  const purchaser = purchaserForCheckout({
    customerId: 7,
    accountEmail: null,
    isAccountVerified: true,
    submittedEmail: 'ada@example.test',
  })

  assert.equal(purchaser.isEmailVerified, false)
})
